File: src/algorithm_java/leetcode/Problem297.java
package algorithm.leetcode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Problem297 {

    // Encodes a tree to a single string.
    public static String serialize(TreeNode root) {
        if (root == null)
            return "";
        Queue<TreeNode> queue = new LinkedList<>();
        List<String> lst = new ArrayList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode cur = queue.poll();
            if (cur != null) {
                lst.add(String.valueOf(cur.val));
            } else {
                lst.add("null");
            }
            if (cur != null) {
                queue.offer(cur.left);
                queue.offer(cur.right);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < lst.size(); i++) {
            sb.append(lst.get(i));
            if (i != lst.size() - 1)
                sb.append(",");
        }
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public static TreeNode deserialize(String data) {
        if (data.length() == 0)
            return null;
        List<TreeNode> nodes = new ArrayList<>();
        String[] ss = data.split(",");
        for (int i = 0; i < ss.length; i++) {
            if (ss[i].equals("null")) {
                nodes.add(null);
            } else {
                nodes.add(new TreeNode(Integer.valueOf(ss[i])));
            }
        }

        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode root = nodes.get(0);
        TreeNode cur = null;
        queue.offer(root);
        int i = 1;
        while (!queue.isEmpty()) {
            cur = queue.poll();
            TreeNode left = nodes.get(i++);
            TreeNode right = nodes.get(i++);
            cur.left = left;
            cur.right = right;
            if (left != null)
                queue.offer(left);
            if (right != null)
                queue.offer(right);
        }
        return root;
    }

    public static void main(String[] args) {
        TreeNode a = new TreeNode(1);
        TreeNode b = new TreeNode(2);
        TreeNode c = new TreeNode(3);
        TreeNode d = new TreeNode(4);
        TreeNode e = new TreeNode(5);
        a.left = b;
        a.right = c;
        c.left = d;
        c.right = e;
        TreeNode r = deserialize("1,2,3,null,null,4,5,null,null,null,null");
        System.out.println();
    }
}

File: src/algorithm_cpp/problem297.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
private:
    vector<string> strsplit(string data, char sp) {
	vector<string> rt;
	int start = 0;
	int len = 0;
	for (int i = 0; i < data.size(); i++) {
	    if (data.at(i) == sp) {
		rt.push_back(data.substr(start, len));
		start = i + 1;
		len = 0;
	    } else {
		len++;
	    }
	}
    rt.push_back(data.substr(start, len));
	return rt;
    }
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
	if (root == nullptr) return "[null]";
	string rt = "[";
        vector<TreeNode*> q;
	q.push_back(root);
	bool qempty = false;
	while(!qempty) {
	    qempty = true;
	    vector<TreeNode*> nextq;
	    for (int i = 0; i < q.size(); i++) {
		if (q[i] == nullptr) {
		    rt += "null";
		    nextq.push_back(nullptr);
		    nextq.push_back(nullptr);
	 	} else {
		    rt += to_string(q[i]->val);
		    nextq.push_back(q[i]->left);
		    nextq.push_back(q[i]->right);
		    if (q[i]->left != nullptr || q[i]->right != nullptr) {
			qempty = false;
		    }
		}
		rt += ",";
	    }
	    q = nextq;
	}
    cout << rt.substr(0, rt.size() - 1) + "]";
	return rt.substr(0, rt.size() - 1) + "]";
		    
	    
    }

    TreeNode* deserializeNode(vector<string> data, int i) {
	if (data[i] == "null" || i >= data.size()) return nullptr;
	TreeNode* node = new TreeNode(stoi(data[i]));
	int leftidx = 2 * i + 1, rightidx = 2 * i + 2;
	node->left = deserializeNode(data, leftidx);
	node->right = deserializeNode(data, rightidx);
	return node;
    }
	
	

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        data = data.substr(1, data.size() - 2);
	vector<string> splits = strsplit(data, ',');
	return deserializeNode(splits, 0);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));
